clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -249,7 +249,14 @@ protected synchronized boolean ensureCoordinatorReady(final Timer timer) {
                 throw fatalException;
             }
             final RequestFuture<Void> future = lookupCoordinator();
-            client.poll(future, timer);
+
+            // if we do not want to block on discovering coordinator at all,
+            // then we should not try to poll in a loop, and should not throw wake-up exception either
+            if (timer.timeoutMs() == 0L) {
```

Comment:
I wonder if this breaks the API in a different way. If the user calls `wakeup()`, would we expect the next blocking call to trigger a wakeup even if it were called with a timeout of zero? I wonder if it would be better to overload `ensureCoordinatorReady` with an additional flag?

Reply:
> would we expect the next blocking call to trigger a wakeup even if it were called with a timeout of zero?

Yes, I think as long as it's a blocking call, the wakeup exception should be thrown, even if it's zero timeout. I've checked existing javadoc in KafkaConsumer, all the methods, which declared wakeupException will be thrown, will still throw wakeupExceptions after this change. So, I think this PR change (1) won't break API, and (2) fixes the issue for commitAsync, which makes sense to me.

> I wonder if it would be better to overload ensureCoordinatorReady with an additional flag?

Yes, as @guozhangwang mentioned in PR description:
> In this PR I'm trying to fix it in a least intrusive way (a more general fix should be, potentially, to have two versions of ensureCoordinatorReady)

I think this fix is safer. But I don't have strong opinion about it, just want to raise the release timing issue here.

Reply:
Yeah, I feel it's a bit slippery to leave the logic in place with just a TODO somewhere to go back and fix it. Too many times, the follow-up never happens. Is it really that much additional effort to add an overload?

```java
protected synchronized boolean ensureCoordinatorReady(final Timer timer) {
  ensureCoordinatorReady(timer, true);
}

private synchronized boolean ensureCoordinatorReady(final Timer timer, boolean checkWakeup) {
...
```